### PR TITLE
Add a new page on "Concurrency and isolates"

### PR DESCRIPTION
**Pull request by ericwindmill**
Adds documentation about using isolates in Flutter apps.

**Staged:** https://flutter-docs-prod--pr9777-isolate-doc-update-ru4mbku4.web.app/perf/isolates

This will address [6564](https://github.com/flutter/website/issues/6564), albeit in a different way than suggested.

### CODE REVIEW

File: src/_data/sidenav.yml
Comment (parlough) [inline src/_data/sidenav.yml:335]: To get the stage workflow running:

```suggestion
      permalink: /data-and-backend/isolates
```

**Comment by domesticmouse** [comment]
Looks like some links got goofed:

```console
check_link_references: Invalid link references found!

data-and-backend/isolates/index.html:
[concurrency page][]
[Dart documentation][]
[concurrency documentation][]
```

**Review by domesticmouse — APPROVED** [review]
LGTM once nits and links are fixed

**Comment by parlough** [comment]
@gaaclarke Would you be willing/interested in reviewing this new page from a technical perspective?

**Comment by ericwindmill** [comment]
Note that @parlough suggested, and I agreed, that this page should live under "Performance and Optimization", rather than Data and Backend. The most recent commits reflect that change

**Comment by parlough** [comment]
@ericwindmill This and your site-www PR are on my todo list to review, but it might not be until Monday. Thanks for working on this exciting articles! I have no doubt users will appreciate the extra insight into concurrency in Dart. 

It's not completely required as I'd be happy to do it as follow-up, but for your own practice and awareness, would you be willing to update both PRs to use semantic breaks. We have some basic details [on site-shared](https://github.com/dart-lang/site-shared/blob/main/doc/writing-for-dart-and-flutter-websites.md#semantic-line-breaks) and there's a more detailed guide at [sembr.org](https://sembr.org/). Essentially they allow for easier editing and review, especially in the future for Git diffs. We don't strictly follow any hard guidelines for it, therefore it's a bit subjective, so don't think too much when adding them, but any progress closer to them helps.

**Comment by ericwindmill** [comment]
> @ericwindmill This and your site-www PR are on my todo list to review, but it might not be until Monday. Thanks for working on this exciting articles! I have no doubt users will appreciate the extra insight into concurrency in Dart.
> 
> It's not completely required as I'd be happy to do it as follow-up, but for your own practice and awareness, would you be willing to update both PRs to use semantic breaks? We have some basic details [on site-shared](https://github.com/dart-lang/site-shared/blob/main/doc/writing-for-dart-and-flutter-websites.md#semantic-line-breaks) and there's a more detailed guide at [sembr.org](https://sembr.org/). Essentially they allow for easier editing and review, especially in the future for Git diffs of changes. We don't strictly follow any hard guidelines for it, therefore it's a bit subjective, so don't think too much when adding them, but any progress closer to them helps.

I'd never heard of this. It's pretty interesting. I'll update the PR, but it probably won't be until Monday.

File: src/_data/sidenav.yml
Comment (sfshaza2) [inline src/_data/sidenav.yml:503]:
```suggestion
    - title: Concurrency and isolates
```

File: src/perf/isolates.md
Comment (sfshaza2) [inline src/perf/isolates.md:2]:
```suggestion
title: Concurrency and isolates
```

File: src/perf/isolates.md
Comment (sfshaza2) [inline src/perf/isolates.md:16]:
```suggestion
that can cause "UI jank" (jerky motion). If your app is experiencing jank for
```

File: src/perf/isolates.md
Comment (sfshaza2) [inline src/perf/isolates.md:18]:
```suggestion
this reason, you can move these computations to a helper isolate. This allows the
underlying runtime environment to run the computation concurrently with the main
```

File: src/perf/isolates.md
Comment (sfshaza2) [inline src/perf/isolates.md:22]:
```suggestion
events in the order that they're added to an event queue. On the main isolate,
```

File: src/perf/isolates.md
Comment (sfshaza2) [inline src/perf/isolates.md:29]:
```suggestion
For smooth rendering, Flutter adds a "paint frame" event to the event queue 60
```

File: src/perf/isolates.md
Comment (sfshaza2) [inline src/perf/isolates.md:53]:
```suggestion
Dart's isolates are an implementation of the [Actor model][]. They can only
```

File: src/perf/isolates.md
Comment (sfshaza2) [inline src/perf/isolates.md:54]:
```suggestion
communicate with each other by message passing. When messages are "passed"
```

File: src/perf/isolates.md
Comment (sfshaza2) [inline src/perf/isolates.md:57]:
```suggestion
mutated on that isolate, doesn't change the value on the original isolate.
```

File: src/perf/isolates.md
Comment (sfshaza2) [inline src/perf/isolates.md:59]: OK, this is a global change - I won't keep calling it out from here on. Change the double and single quotes to the straight up and down kind.  :)

```suggestion
The only objects that aren't copied when passed to an isolate are immutable
```

File: src/perf/isolates.md
Comment (sfshaza2) [inline src/perf/isolates.md:66]:
```suggestion
the `Isolate.exit` method. Because the sending isolate won't exist after
```

File: src/perf/isolates.md
Comment (sfshaza2) [inline src/perf/isolates.md:72]: Avoid future tense.

```suggestion
makes a copy of a mutable message as it sends, and `Isolate.exit`, which
sends the reference to the message. Both `Isolate.run` and `compute`
```

File: src/perf/isolates.md
Comment (sfshaza2) [inline src/perf/isolates.md:123]:
```suggestion
your Flutter app, you should use the `compute` method. The [`compute()`][] method
```

File: src/perf/isolates.md
Comment (sfshaza2) [inline src/perf/isolates.md:129]:
```suggestion
Please add a sentence introducing this list. Such as "The following list describes longer-lived isolates:"
```

File: src/perf/isolates.md
Comment (sfshaza2) [inline src/perf/isolates.md:152]: Generally, if we uppercase the class and are talking about a specific class, we code font it. If we don't uppercase it, and are using the word generically, we don't code font it.

```suggestion
`Ports` behave similarly to `Streams`, in which the `StreamController` or `Sink` is
```

File: src/perf/isolates.md
Comment (sfshaza2) [inline src/perf/isolates.md:155]:
```suggestion
analogy, the `StreamConroller` is called a `SendPort`, and you can "add" messages
with the `send()` method. `ReceivePort`s are the listeners, and when these listeners
```

File: src/perf/isolates.md
Comment (sfshaza2) [inline src/perf/isolates.md:183]: I can't help but think that this very useful list should appear MUCH higher in this page.

File: src/perf/isolates.md
Comment (sfshaza2) [inline src/perf/isolates.md:194]:
```suggestion
Platform channel isolates use the [`BackgroundIsolateBinaryMessenger`][] API. The
```

File: src/perf/isolates.md
Comment (sfshaza2) [inline src/perf/isolates.md:248]:
```suggestion
can't access assets using `rootBundle` in spawned isolates, nor can you perform
```

File: src/perf/isolates.md
Comment (sfshaza2) [inline src/perf/isolates.md:253]:
```suggestion
With background isolate platform channels, you can use platform channels in
```

File: src/perf/isolates.md
Comment (sfshaza2) [inline src/perf/isolates.md:271]:
```suggestion
- Read more about the `BackgroundIsolateBinaryMessenger` API [announcement][].
```

File: src/perf/isolates.md
Comment (sfshaza2) [inline src/perf/isolates.md:285]:
```suggestion
[`BackgroundIsolateBinaryMessenger`]: {{site.api}}/flutter/services/BackgroundIsolateBinaryMessenger-class.html
```

**Review by sfshaza2 — COMMENTED** [review]
Overall, this looks really good. There are numerous nits, many that involve code fonting, and the single and double quotes that are morphed to the "curly" style.

**Comment by ericwindmill** [comment]
> I'd never heard of this. It's pretty interesting. I'll update the PR, but it probably won't be until Monday.

@parlough - I added semantic breaks. I'm not confident that I did it 100% correctly, so feel free to let me know if there are any issues.

File: examples/development/concurrency/isolates/lib/main.dart
Comment (navaronbracke) [inline examples/development/concurrency/isolates/lib/main.dart:40]: Can we improve the types here a bit? Primarily giving `Isolate.run()` its type argument,
and using a cast<T>() for the List, which makes it a bit cleaner. I do not feel strongly about changing dynamic -> `Object?`

```suggestion
Future<List<Photo>> getPhotos() async {
  final String jsonString = await rootBundle.loadString('assets/photos.json');

  final List<Photo> photos = await Isolate.run<List<Photo>>(() {
    // jsonDecode() gives untyped lists.
    final List<Object?> photoData = jsonDecode(jsonString) as List<Object?>;
    // So cast here, and then you could use a tear-off as well.
    return photoData.cast<Map<String, Object?>>().map(Photo.fromJson).toList();
  });

  return photos;
}
```

File: src/perf/isolates.md
Comment (navaronbracke) [inline src/perf/isolates.md:38]: I only saw 3 green events in the image. Unless you were referring to the blue rectangles?

File: src/perf/isolates.md
Comment (navaronbracke) [inline src/perf/isolates.md:185]: Having to recommend `compute()` on the web, where it runs things on the main thread anyway, seems kind of defeating the point of offloading work onto a different thread/isolate.

Instead, you can refer to the `WebWorker` API, which is actually quite similar to the Isolate API (start/stop a WebWorker, passing messages & transferable by-reference data types).

We have first-party support in [package:web](https://github.com/dart-lang/web/blob/e99d53716709d1cbce5533a7f6d036a8a5e19f68/lib/src/dom/html.dart#L4782) You could also link to https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API

Perhaps a code sample where you write the photo sample with a WebWorker could be nice to have?

**Review by navaronbracke — COMMENTED** [review]
Some minor feedback while skimming through it.

File: src/perf/isolates.md
Comment (ericwindmill) [reply src/perf/isolates.md:185]: Overwhelmingly, the most requested addition to our docs about isolates is being clear about the limitations. We have resources sprinkled across the web that use `compute` in example code, and it's important that we're explicit about what that means. 

I think documenting WebWorkers is beyond the scope of this update.

File: examples/development/concurrency/isolates/lib/main.dart
Comment (ericwindmill) [reply examples/development/concurrency/isolates/lib/main.dart:40]: I'm adding this suggested update to my next commit, as I'll need to update the snippets in the markdown docs that this uses as well.

File: src/perf/isolates.md
Comment (parlough) [reply src/perf/isolates.md:185]: I agree we don't need to document web workers here, but we do need to be careful about how we write about `compute` and its behavior on web.

Some people incorrectly see it as "using `compute` instead of `Isolate.run` allows the same behavior to work on web".

I wonder if we should just always talk about and default to `Isolate.run`, but then in a small note or similar put forward `compute` as a way that allows the same code to **compile** for web without changes, but highlight that it does **not work** the same. That way it's a more explicit choice they have to make after they know the differences.

File: .github/workflows/build-and-links.yml
Comment (johnpryan) [inline .github/workflows/build-and-links.yml:20]: Is this change intentional?

File: .github/workflows/stage.yml
Comment (johnpryan) [inline .github/workflows/stage.yml:35]: Same question, are you intending to make changes to the GitHub actions here?

File: src/_data/catalog/index.json
Comment (johnpryan) [inline src/_data/catalog/index.json:67]: Is this intentional? It doesn't seem related to isolates.

File: .github/workflows/build-and-links.yml
Comment (parlough) [reply .github/workflows/build-and-links.yml:20]: It seems like this and the other changes were brought in from `main` by accident with a merge.

File: src/_data/catalog/index.json
Comment (ericwindmill) [reply src/_data/catalog/index.json:67]: no, I don't know how this happened. I will revert

File: src/perf/isolates.md
Comment (gaaclarke) [inline src/perf/isolates.md:38]: nit: These lines are *aggressively* newline'd.  You might find it helpful to use plugins for vscode that will automatically do the wordwrap for you.  I use this one: https://marketplace.visualstudio.com/items?itemName=stkb.rewrap

File: src/perf/isolates.md
Comment (gaaclarke) [inline src/perf/isolates.md:53]: I would remove the "or" since it isn't clear you are describing the frame gap like this.

File: src/perf/isolates.md
Comment (gaaclarke) [inline src/perf/isolates.md:110]: Can you link to the Dart documentation on these rules?

File: src/perf/isolates.md
Comment (gaaclarke) [inline src/perf/isolates.md:112]: "Transferred" is ambiguous, I'd say that a reference to the data is sent.

File: src/perf/isolates.md
Comment (gaaclarke) [inline src/perf/isolates.md:100]: Maybe it's worth mentioning "port"s here?

I think in general in this doc whenever it's using the "passing messages" nomenclature it might be more clear to use Dart nomenclature like "ports".

File: src/perf/isolates.md
Comment (gaaclarke) [inline src/perf/isolates.md:120]: I'd say it passes ownership of the message from one isolate to another.

File: src/perf/isolates.md
Comment (gaaclarke) [inline src/perf/isolates.md:132]: It may be worth mentioning the overhead associated copying messages and spawning isolates so people get some intuition about the tradeoffs.

File: src/perf/isolates.md
Comment (gaaclarke) [inline src/perf/isolates.md:223]: FWIW I have an easier way https://medium.com/@gaaclarke/isolate-agents-easy-isolates-for-flutter-6d75bf69a2e7.

File: src/perf/isolates.md
Comment (gaaclarke) [inline src/perf/isolates.md:238]:
```suggestion
the `shared_preferences` package in a background isolate.
```

File: src/perf/isolates.md
Comment (gaaclarke) [inline src/perf/isolates.md:270]: The nomenclature of this document is "message passing", I wonder if it would be better to use the nomenclature of Dart which is "ports".

**Review by gaaclarke — COMMENTED** [review]
LGTM, I think people are going to find this really helpful.  I just have clarity suggestions here and there, nothing was factually wrong.

File: src/perf/isolates.md
Comment (ericwindmill) [reply src/perf/isolates.md:38]: Yeah.. this is a practice called [Semantic Line Breaks](https://sembr.org/) that I was introduced to via previous comments on this very PR. I'm hoping someone with more knowledge about "Sembr" can tell me if I did it _too aggressively_.

**Comment by ericwindmill** [comment]
@Sfshaza PTAL

**Comment by parlough** [comment]
Thanks for all of your work on this @ericwindmill! It's really exciting to have fleshed out docs about these important topics.

I'm likely not going to have enough time to review this PR this week, but it looks like you've got a good amount of feedback from the right people, so I'll remove my assignment. I'll defer to @atsansone and/or @sfshaza2 to do a final copy pass after the changes you made and then they/you can merge it.

I will make sure to take a look once I have time though and let you know if I have any further thoughts :)

**Review by sfshaza2 — APPROVED** [review]
A few very minor nits. Overall, this looks fantastic, @ericwindmill! Once you fix the minor nits, let's land this puppy! (I'll go ahead and approve it to make things easier.)

**Comment by sfshaza2** [comment]
I'm landing this! Any further changes (I'm not aware of any) can be handled in a new PR. Thanks, @ericwindmill!

**Pull request merged**
